### PR TITLE
Shallow cloning functionality for the `git://` file handler

### PR DESCRIPTION
**Pull request by Wuestengecko**
This PR implements the capability for the `git://` file handler to create and update shallow clones in its cache directory. This significantly reduces the network bandwidth and disk space needed for updating and maintaining the cache.

It also introduces the new file handler specific keyword argument `shallow=`, which controls this behavior. If it is set to `True` (the default), clones and fetches will be shallow, and existing full clones will be converted to shallow clones as they are accessed. If it is `False`, clones and fetches will fetch the full history. Existing shallow clones will however not be "unshallowed".

Note that actually unlinking unused objects is subject to git's garbage collection. To benefit from the disk space savings immediately, consider either wiping the cache or updating all locally cached models (by loading them once) and running `git gc` manually.

- [x] When creating a new cache, the relevant ref is downloaded and made available as symbolic ref.
- [x] When updating a cache, the local copy of the ref is updated to the current state on the remote.
- [x] When pushing to a branch, the remote is updated with the committed changes.

### CODE REVIEW

**Review by ewuerger — CHANGES_REQUESTED** [review]
Very good work as always. Since you have more knowledge about git than me I don't have any doubts about the newly used git commands. What I'd advise though is that you are working on the documentation of the `GitFileHandler`. The attributes (apart from username and password) can have a docstring. The `open` method doesn't have a docstring yet and `write_transaction` is referencing a private class' docstring (`_GitTransaction`).

Some points that could need improvement. You can decide if these changes are to be done in this PR also or in another future one. You know my stance on this: The best time to improve sth. is now. 

The only thing that I'm truly missing are tests that prove your claims:
As a starting point I've added a test called [`test_shallow_cloning_uses_less_space_than_full_clone`](https://github.com/DSD-DBS/py-capellambse/pull/99/files#diff-11e65aea2350d555efc69bcc3004a2cab7e8cb52264f18901492bbba6150d323R40-R57)
which shows that an unshallow cloned cache directory actually has the same size compared to a shallow cloned cache directory.

File: capellambse/loader/filehandler/gitfilehandler.py [hunk 1]
```diff
@@ -641,13 +648,45 @@ def __init_cache_dir_remote(self) -> None:
...
+                fetchspec = f"+{self.revision}"
+                if not git_object_name.search(self.revision):
+                    fetchspec += f":{self.revision}"
+                shallow_opts = ("--depth=1", fetchspec)
+            self._git("fetch", self.path, *shallow_opts)
+
+    def __resolve_remote_ref(self, ref: str) -> tuple[str, str]:
+        """Resolve the given ``ref`` on the remote."""
+        LOGGER.debug("Resolving ref %r on remote %s", ref, self.path)
+        listing = self._git("ls-remote", self.path, ref, encoding="ascii")
```
[the remark this answers is not in the document]
Comment (Wuestengecko) [reply capellambse/loader/filehandler/gitfilehandler.py:672]: Good catch! For some reason I was under the impression that branch names could only contain ASCII, but while it's probably a good practice (just like with filenames ;) ), this is not enforced.

File: capellambse/loader/filehandler/gitfilehandler.py [hunk 2]
```diff
@@ -35,6 +35,8 @@
 
 LOGGER = logging.getLogger(__name__)
 
+git_object_name = re.compile("(^|/)([0-9a-fA-F]{4,}|(.+_)?HEAD)$")
```
[the remark this answers is not in the document]
Comment (Wuestengecko) [reply capellambse/loader/filehandler/gitfilehandler.py:38]: The public-ness is a leftover from development, it should have been private from the start. Gonna fix that.

File: tests/test_model_loading.py [hunk 3]
```diff
@@ -26,11 +27,14 @@ def test_model_loading_via_LocalFileHandler(path: str | pathlib.Path):
     capellambse.MelodyModel(path)
 
 
-def test_model_loading_via_GitFileHandler():
-    path = "git+" + pathlib.Path.cwd().as_uri()
-    capellambse.MelodyModel(
-        path, entrypoint="tests/data/melodymodel/5_0/Melody Model Test.aird"
-    )
+class TestGitFileHandler:
```
Comment (ewuerger) [inline tests/test_model_loading.py:30]: > On one of our production repositories, this reduces the amount of data
transferred by about 90%, and the time until the `MelodyModel` object
becomes ready during an initial clone by roughly the same amount.

I'm convinced that the network-bandwidth reduction happens b/c I trust in git. But I can remember that I checked your claim regarding space benefits and couldn't prove it. Do you want to prove it with either a showcase or a dedicated test case right here? I promise this is the last time I raise this.

**Review by ewuerger — CHANGES_REQUESTED** [review]
Beside missing proof of gained space claim: :star2:  LGTM :star2:.

File: tests/test_model_loading.py (hunk 3 above)
Comment (ewuerger) [reply tests/test_model_loading.py]: I sense dark magic here. You can pass.

**Pull request merged**